src/param_env.md

```diff
@@ -1,69 +1,160 @@
+
+[elaborate]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_infer/traits/util/fn.elaborate.html
+[next-gen-solver]: https://rustc-dev-guide.rust-lang.org/solve/trait-solving.html
+
+### Normalizing all bounds
+
+In the old trait solver the where clauses stored in `ParamEnv` are required to be fully normalized or else the trait solver will not function correctly.  A concrete example of needing to normalize the `ParamEnv` is the following:
+```rust
+trait Trait<T> {
     type Assoc;
 }
 
-trait Bar { }
+trait Other {
+    type Bar;
+}
 
-trait Baz {
-    fn stuff() -> bool;
+impl<T> Other for T {
+    type Bar = u32;
 }
 
-fn foo<T>(t: T)
+// `foo`'s unnormalized `ParamEnv` would be:
+// `[T: Sized, U: Sized, U: Trait<T::Bar>]`
+fn foo<T, U>(a: U) 
 where
-    T: Foo,
-    <T as Foo>::Assoc: Bar
+    U: Trait<<T as Other>::Bar>,
 {
-   bar::<T::Assoc>()
+    requires_impl(a);
 }
 
-fn bar<T: Baz>() {
-    if T::stuff() { mep() } else { mop() }
-}
+fn requires_impl<U: Trait<u32>>(_: U) {}
 ```
 
-We may know some things inside `bar` that we wouldn't know if we just fetched
-`bar`'s param env because of the `<T as Foo>::Assoc: Bar` bound in `foo`. This
-is a contrived example that makes no sense in our existing analyses, but we may
-run into similar cases when doing analyses with associated constants on generic
-traits or traits with assoc types.
+As humans we can tell that `<T as Other>::Bar` is equal to `u32` so the trait bound on `U` is equivalent to `U: Trait<u32>`. In practice trying to prove `U: Trait<u32>` in the old solver in this environment would fail as it is unable to determine that `<T as Other>::Bar` is equal to `u32`.
+
+To work around this we normalize `ParamEnv`'s after constructing them so that `foo`'s `ParamEnv` is actually: `[T: Sized, U: Sized, U: Trait<u32>]` which means the trait solver is now able to use the `U: Trait<u32>` in the `ParamEnv` to determine that the trait bound `U: Trait<u32>` holds.
+
+This workaround does not work in all cases as normalizing associated types requires a `ParamEnv` which introduces a bootstrapping problem. We need a normalized `ParamEnv` in order for normalization to give correct results, but we need to normalize to get that `ParamEnv`. Currently we normalize the `ParamEnv` once using the unnormalized param env and it tends to give okay results in practice even though there are some examples where this breaks ([example]).
+
+In the next-gen trait solver the requirement for all where clauses in the `ParamEnv` to be fully normalized is not present and so we do not normalize when constructing `ParamEnv`s.
+
+[example]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e6933265ea3e84eaa47019465739992c
+
+---
+
+When needing a `ParamEnv` in the compiler there generally three options for obtaining one:
+- The correct env is already in scope simply use it (or pass it down the call stack to where you are)
```

Comment:
Maybe explain where you're likely to find it: in `FnCtxt`, in `LateLintContext`, anywhere else? Maybe `WfCheckingCtxt`?